example3/OpenTelemetry.Metric.Sdk/MetricProvider.cs

```diff
@@ -0,0 +1,396 @@
+using System.Threading.Tasks;
+using System.Threading;
+using System.Net;
+using Microsoft.Diagnostics.Metric;
+using OpenTelemetry.Metric.Api;
+
+namespace OpenTelemetry.Metric.Sdk
+{
+    public class MetricProvider
+    {
+        private readonly object lockMeters = new();
+        private List<Meter> meters = new();
+
+        private ConcurrentDictionary<AggregatorKey, AggregatorState> aggregateDict = new();
+        private Action<MeterBase, double, AggregatorKey> _updateAggregatorByKeyFunc;
+
+        private string name;
+        private int collectPeriod_ms = 2000;
+        private bool isBuilt = false;
+        private bool flushCollect = false;
+        private CancellationTokenSource cts = new();
+
+        private List<Tuple<string,string>> metricFilterList = new();
+
+        private List<(Aggregator agg, MetricLabelSet[] labels)> aggregateByLabelSet = new();
+
+        private List<Exporter> exporters = new();
+
+        private Task collectTask;
+        private Task dequeueTask;
+
+        private MeterListener listener;
+
+        private ConcurrentQueue<Tuple<MeterBase,double,string[],object>> incomingQueue = new();
+        private bool useQueue = false;
+
+        public MetricProvider()
+        {
+            this.listener = new MeterListener()
+            {
+                MeterPublished = OnMeterPublished,
+                MeasurementRecorded = OnMeasurementRecorded
+            };
+            // cache this delegate so we don't keep allocating it
+            this._updateAggregatorByKeyFunc = this.UpdateAggregatorByKey;
+        }
+
+        public MetricProvider Name(string name)
+        {
+            this.name = name;
+            return this;
+        }
+
+        public MetricProvider Include(string term)
+        {
+            metricFilterList.Add(Tuple.Create("Include", term));
+            return this;
+        }
+
+        public MetricProvider AddMetricExclusion(string term)
+        {
+            metricFilterList.Add(Tuple.Create("Exclude", term));
+            return this;
+        }
+
+        public MetricProvider AggregateByLabels(Aggregator agg, params MetricLabelSet[] labelset)
+        {
+            aggregateByLabelSet.Add((agg, labelset));
+            return this;
+        }
+
+        public MetricProvider AddExporter(Exporter exporter)
+        {
+            exporters.Add(exporter);
+            return this;
+        }
+
+        public MetricProvider SetCollectionPeriod(int milliseconds)
+        {
+            collectPeriod_ms = milliseconds;
+            return this;
+        }
+
+        public MetricProvider UseQueue()
+        {
+            useQueue = true;
+            return this;
+        }
+
+        public MetricProvider Build()
+        {
+            // Start Periodic Collection Task
+
+            var token = cts.Token;
+
+            collectTask = Task.Run(async () => {
+                while (!flushCollect && !token.IsCancellationRequested)

```

Comment:
I read this as: Metrics is "push" based similar to traces..
In the old-but-now-deleted-otel-metric-sdk spec, there was the notion of a Controller, which did the task of pushing the metrics on a schedule, instead of the provider pushing it out to exporters. 
Just wondering if there was any discussion around that model vs push model as in this PR?